### ISSUE TITLE
HAProxy 2.8.5 setting very large window updates with HTTP2

### ISSUE THREAD

**Issue by bnesbitt** [haproxy/haproxy#2447]
### Detailed Description of the Problem

We have observed HAProxy setting very large TCP window updates in our layer 7 health checks. When this happens we have observed that the health check can time out. The following is a snap shot of one of the health checks:

![haproxy_pcap](https://github.com/haproxy/haproxy/assets/2341705/f1db2422-f59f-4398-be4e-86919965b54c)

The health check is a HTTP2 HEAD request:
`http-check send meth GET uri http://%s/oam/v1/operational_status?for_georegion=geoRegion1 ver HTTP/2 hdr xxx-health-check true`

At times this can result in a huge window update followed by a RST that seems to then result in the health check failing
![window_update](https://github.com/haproxy/haproxy/assets/2341705/58e5aa13-736f-4873-9b40-635d44cee605)

One of these window size increments in the pair is usually set to the maximum allowed HTTP2 window size value (2^31).

What could be causing the window size to be set to such a large value, and why would HAProxy send a RST immediately afterwards? The health check timeout is 1 second, and this hasn't been breached in this example.

When we run load through HAProxy with this health check then after a few minutes the window updates can occur, then health checks start failing. **However**, if we disable the layer 7 health checks and drop down to layer 4 checks, then we don't observe this problem.

### Expected Behavior

We're not sure what the behaviour should be, or why the window update is so big.

Is this being set explicitly by HAProxy itself, or by the kernel, and can this be clamped by using either of these?
`net.ipv4.tcp_rmem`
`net.core.rmem_max`

### Steps to Reproduce the Behavior

We can only reproduce this at a customer lab, and cannot reproduce it internally which is hindering our investigation. There could well be congestion within the customer lab.

The scenario is that we are loading testing HTTP2 with HAProxy balancing the requests across 40 backend servers. With the layer 7 health checks we end up with huge window updates followed by RST that fails the health checks and causes timeouts in the backends. With layer 4 only health checks we do not observe this behaviour.

### Do you have any idea what may have caused this?

Not yet.

### Do you have an idea how to solve the issue?

We are investigating if we can use layer 4 checks only as a workaround, but this isn't ideal.

### What is your configuration?

```haproxy
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# #
# #   HAProxy Ingress Controller
# #   --------------------------
# #   This file is automatically updated, do not edit
# #
#
global
    master-worker
    unix-bind mode 0600
    stats socket /var/run/haproxy/admin.sock level admin expose-fd listeners mode 600
    maxconn 100000
    hard-stop-after 10m
    log 127.0.0.1:514 len 1024 format rfc5424 local0
    log-tag ingress
    lua-load /etc/haproxy/lua/services.lua
    lua-load /etc/haproxy/lua/responses.lua
    tune.ssl.default-dh-param 2048
    ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
    ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
    ssl-default-bind-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets
    ssl-default-server-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
    ssl-default-server-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
    lua-prepend-path /etc/REDACTED/?.lua
    lua-load /etc/REDACTED/XXX.lua
    tune.h2.max-concurrent-streams 1000
    tune.h2.initial-window-size 1048576
    tune.bufsize 131072
    nbthread 1

defaults
    log global
    maxconn 100000
    option redispatch
    option dontlognull
    option http-server-close
    option http-keep-alive
    timeout client          50s
    timeout client-fin      50s
    timeout connect         5s
    timeout http-keep-alive 1m
    timeout http-request    5s
    timeout queue           5s
    timeout server          50s
    timeout server-fin      50s
    timeout tunnel          1h
    option http-no-delay
    timeout connect 200
    timeout client 300s
    timeout server 5000
    option dontlog-normal
    no option prefer-last-server
    timeout client-fin 1s
    timeout server-fin 1s
    no option http-server-close

  # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# #
# #   BACKENDS
# #
#
backend XXX_west-site1-ingress-local_3480
    mode http
    balance roundrobin
    acl https-request ssl_fc
    http-request set-header X-Original-Forwarded-For %[hdr(x-forwarded-for)] if { hdr(x-forwarded-for) -m found }
    http-request del-header x-forwarded-for
    option forwardfor
    http-reuse always
    option nolinger
    option httpchk
    http-check connect proto h2
    http-check send meth HEAD uri http://%s/oam/v1/operational_status?for_georegion=geoRegion1 ver HTTP/2 hdr XXX-health-check true
    http-check expect rstatus [0-9]+
    default-server max-reuse 500 fall 1 rise 5
    retries 3
    retry-on conn-failure empty-response
    option redispatch
    http-response set-header Strict-Transport-Security "max-age=15768000" if https-request
    server srv001 10.244.108.83:3480 weight 1 proto h2 maxconn 5000 check inter 1s
    server srv002 10.244.229.110:3480 weight 1 proto h2 maxconn 5000 check inter 1s
    server srv003 10.244.229.89:3480 weight 1 proto h2 maxconn 5000 check inter 1s
    server srv004 10.244.108.74:3480 weight 1 proto h2 maxconn 5000 check inter 1s
backend XXX_west-site1-XXX-haproxy-ingress-no-partition_http
    mode http
    balance roundrobin
    acl https-request ssl_fc
    http-request set-header X-Original-Forwarded-For %[hdr(x-forwarded-for)] if { hdr(x-forwarded-for) -m found }
    http-request del-header x-forwarded-for
    option forwardfor
    http-request set-var(req.client_ip) fc_src  # The client IP address.
    http-request set-var(req.url) url           # Store the full URL.
    http-request use-service lua.partition_not_found
    http-response set-header Strict-Transport-Security "max-age=15768000" if https-request
    server srv001 127.0.0.1:1023 disabled weight 1 proto h2 maxconn 5000 check inter 1s
backend XXX_west-site2-ingress-remote_80
    mode http
    balance roundrobin
    acl https-request ssl_fc
    http-request set-header X-Original-Forwarded-For %[hdr(x-forwarded-for)] if { hdr(x-forwarded-for) -m found }
    http-request del-header x-forwarded-for
    option forwardfor
    http-reuse always
    option nolinger
    option httpchk
    http-check connect proto h2
    http-check send meth GET uri http://%s/oam/v1/operational_status?for_georegion=geoRegion1 ver HTTP/2 hdr XXX-health-check true
    http-check expect status 200
    default-server max-reuse 500 fall 1 rise 5
    retries 3
    retry-on conn-failure empty-response
    http-response set-header Strict-Transport-Security "max-age=15768000" if https-request
    server srv001 192.168.230.14:80 weight 1 proto h2 maxconn 5000 check inter 1s
backend XXX_west-site3-ingress-remote_80
    mode http
    balance roundrobin
    acl https-request ssl_fc
    http-request set-header X-Original-Forwarded-For %[hdr(x-forwarded-for)] if { hdr(x-forwarded-for) -m found }
    http-request del-header x-forwarded-for
    option forwardfor
    http-reuse always
    option nolinger
    option httpchk
    http-check connect proto h2
    http-check send meth GET uri http://%s/oam/v1/operational_status?for_georegion=geoRegion1 ver HTTP/2 hdr XXX-health-check true
    http-check expect status 200
    default-server max-reuse 500 fall 1 rise 5
    retries 3
    retry-on conn-failure empty-response
    http-response set-header Strict-Transport-Security "max-age=15768000" if https-request
    server srv001 192.168.231.10:80 weight 1 proto h2 maxconn 5000 check inter 1s


  # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# #
# #   FRONTENDS
# #
#

  # # # # # # # # # # # # # # # # # # #
# #
#     HTTP frontend
#
frontend _front_http
    mode http
    bind :80
    option httplog
    # REDACTED: Begin
    # Define the ACLs first.
    acl has_imsi_in_url url_reg -i imsi-\d+
    acl check_body_for_imsi path_reg -i subs-to-notify
    acl check_body_for_imsi path_reg -i \/nudr-dr\/v.*\/application-data
    
    acl is_udsf path -i -m beg /nudsf-dr
    acl is_roaming path -i -m reg /roaming-plan-profile
    acl no_XXX_partition_uuid hdr_cnt(XXX-partition-uuid) eq 0
    acl no_XXX_start hdr_cnt(XXX-ingress-start) eq 0
    
    # ACL to detect ILD health checks. We will use this to send to the local backend.
    acl is_georegion_health_check path -i -m beg /oam/v1/operational_status
    acl is_operational_status path -i -m beg /builtin/v1/operationalstatus
    
    # We only buffer the request if we know what type of request it is.
    http-request wait-for-body time 5s at-least 16k if check_body_for_imsi
    http-request set-var(req.body) req.body if check_body_for_imsi
    
    # Store data that we need for the Lua scripts.
    http-request set-var(req.client_ip) fc_src        # The client IP address.
    http-request set-var(req.client_port) fc_src_port # The client port.
    http-request set-var(req.url) url                 # Store the full URL.
    http-request set-var(req.path) path               # The URL path.
    http-request set-var(req.method) method           # The HTTP method - to detect DELETE requests.
    http-request set-var(req.existing_uuid) req.hdr(XXX-partition-uuid)
    
    # Response variables: Begin
    # We use these for logging and diagnostics purposes.
    http-response set-var(res.client_ip) fc_src
    http-response set-var(res.client_port) fc_src_port
    http-response set-var(res.method) method
    http-response set-var(res.uuid) res.hdr(XXX-partition-uuid)
    http-response set-var(res.health_check) res.hdr(XXX-health-check)
    http-response set-var(res.start_time) res.hdr(XXX-ingress-start)
    # Response variables: End
    
    http-request set-var(req.partition_trace_enabled) req.hdr(XXX-specialattributes)
    
    # Record where we came from if sending from one haproxy to another:
    http-request set-var(req.proxy_origin) req.hdr(XXX-partition-origin)
    
    # Assign a unique id to the transaction and add it to the request.
    http-request set-var(req.uuid) uuid() if no_XXX_partition_uuid
    # Only add the UUID header if it doesn't exist. It may already have been set by an upstream
    # HAProxy which we must preserve.
    http-request add-header XXX-partition-uuid %[var(req.uuid)] if no_XXX_partition_uuid
    
    # Add the epoch timestamp (ms) to the request headers. Only add this header if
    # it hasn't already been set by an upstream HAProxy.
    http-request add-header XXX-ingress-start %[date(0,us)] if no_XXX_start
    
    # Used to determine if requests should be routed only to the local partition.
    http-request set-var(req.local_partition_only) str("true") if is_udsf or is_roaming or is_georegion_health_check or is_operational_status
    http-request set-var(req.is_udsf) str("true") if is_udsf
    http-request set-var(req.is_health_check) str("true") if is_georegion_health_check
    
    http-request lua.get_imsi_from_url if has_imsi_in_url
    http-request lua.scan_request_for_imsi if check_body_for_imsi
    
    # Response handler
    http-response lua.handle_response
    
    use_backend %[lua.select_partition]
    # REDACTED: End
    http-request set-var(req.path) path
    http-request set-var(req.host) hdr(host),field(1,:),lower
    http-request set-var(req.base) var(req.host),concat(\#,req.path)
    http-request set-header X-Forwarded-Proto http
    http-request del-header X-SSL-Client-CN
    http-request del-header X-SSL-Client-DN
    http-request del-header X-SSL-Client-SHA1
    http-request del-header X-SSL-Client-SHA2
    http-request del-header X-SSL-Client-Cert
    http-request set-var(req.defaultbackend) str(<default>\#),concat(,req.path),map_str(/etc/haproxy/maps/_front_defaulthost__exact.map) if !{ var(req.backend) -m found }
    use_backend %[var(req.backend)] if { var(req.backend) -m found }
    use_backend %[var(req.defaultbackend)]
    default_backend XXX_west-site1-XXX-haproxy-ingress-no-partition_http


  # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# #
# #   SUPPORT
# #
#

  # # # # # # # # # # # # # # # # # # #
# #
#     Stats
#
listen stats
    mode http
    bind :1936
    stats enable
    stats uri /
    no log
    option httpclose
    stats show-legends

  # # # # # # # # # # # # # # # # # # #
# #
#     Prometheus metrics
#
frontend prometheus
    mode http
    bind :9101
    http-request use-service prometheus-exporter if { path /metrics }
    http-request use-service lua.send-prometheus-root if { path / }
    http-request use-service lua.send-404
    no log

  # # # # # # # # # # # # # # # # # # #
# #
#     Monitor URI
#
frontend healthz
    mode http
    bind :10253
    monitor-uri /healthz
    http-request use-service lua.send-404
    no log
```

### Output of `haproxy -vv`

```plain
# haproxy -vv
HAProxy version 2.8.5-aaba8d0 2023/12/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2028.
Known bugs: http://www.haproxy.org/bugs/bugs-2.8.5.html
Running on: Linux 4.18.0-348.12.2.el8_5.x86_64 #1 SMP Mon Jan 17 07:06:06 EST 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=12).
Built with OpenSSL version : OpenSSL 1.1.1w  11 Sep 2023
Running on OpenSSL version : OpenSSL 1.1.1w  11 Sep 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20210110

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

### Additional Information

We deploy HAProxy as K8s as an ingress controller using the following HAProxy image:
**image: haproxy:2.8.5**

**Comment by bnesbitt** [comment]
One further point is that we're using lots of haproxy containers that are balanced by the k8s ingress controller meaning they are all sending lots of health checks. 

Is there a way to keep the health check connections open instead of closing them with an RST every time? Seems like https://github.com/haproxy/haproxy/issues/560 may be somewhat related?

**Comment by wtarreau** [comment]
The window update is normal and is distinct from the RST. At the moment it's sent, it's not yet known that the connection will be abandoned, that's why it's sent.

The question should rather be "why are we sending an RST" in fact, and the RST is sent at the end of a health check to release the connection once the response arrives. Since HTTP checks were rewritten, there were some discussions about offering the ability to reuse existing connections. This should never be done by default though, because it doesn't validate that the server can be reached. One classical problem on some servers is that they're running a sub-optimal conntrack value and will stop accepting connections past a certain number. At least with checks you discover the problem. When reusing connections, you believe that the server is sill usable but new connections fail.

The problem you're reporting (failing health checks) should be addressed however. There is zero reason such a check should fail, and it's very possible that an error is detected, or something like this. You're speaking about a check timeout, what do you see exactly in your logs or check reports ? If you can reproduce it easily on an idle server, it would be really nice to enable H2 traces so that we try to figure what happens. I really suspect that the connection closure is mistakenly detected as an error but then it should be seen as an instant failure and not as a timeout. That's why I'm interested in the details!

**Comment by bnesbitt** [comment]
Thanks for your fast and detailed response.

After further investigation we think the issue lies with the backend servers. It seems that they are not closing connections properly when the health checks send a RST. We managed to get around this by changing the health check to:

`http-check connect proto h2 linger`

Where we added the **linger** option. With this we can see the health check terminate with a fin-ack. We understand that this isn't as efficient as a RST, but we will have to live with that until we fix the RST handling on our backend servers.

I think we can close this one.
